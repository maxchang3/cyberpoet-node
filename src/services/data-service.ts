import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import type { PoemLine, PoetryConfig, SentenceStructure, SpecialWord, WordRecord, WorkingStructure } from '@/types'

/**
 * 数据访问层 - 负责加载和管理诗歌生成所需的数据
 */
export class DataService {
    private static instance: DataService
    private dataDir: string
    private configDir: string

    // 缓存的数据
    private config: PoetryConfig | null = null
    private nouns: WordRecord[] | null = null
    private adjectives: WordRecord[] | null = null
    private intransitiveVerbs: WordRecord[] | null = null
    private transitiveVerbs: WordRecord[] | null = null
    private interjections: WordRecord[] | null = null
    private sentenceStructures: SentenceStructure[] | null = null
    private workingStructures: WorkingStructure[] | null = null
    private specialWords: SpecialWord[] | null = null
    private poemLines: PoemLine[] | null = null

    private constructor() {
        this.dataDir = join(process.cwd(), 'src', 'data')
        this.configDir = process.cwd()
    }

    /** 获取单例实例 */
    public static getInstance(): DataService {
        if (!DataService.instance) {
            DataService.instance = new DataService()
        }
        return DataService.instance
    }

    /** 加载JSON文件 */
    private loadJsonFile<T>(filename: string): T {
        try {
            const filePath = join(this.dataDir, filename)
            const content = readFileSync(filePath, 'utf-8')
            return JSON.parse(content) as T
        } catch (error) {
            throw new Error(`无法加载数据文件 ${filename}: ${error}`)
        }
    }

    /** 加载配置文件 */
    private loadConfigFile(): PoetryConfig {
        try {
            const configPath = join(this.configDir, 'config.json')
            const content = readFileSync(configPath, 'utf-8')
            return JSON.parse(content) as PoetryConfig
        } catch (error) {
            throw new Error(`无法加载配置文件: ${error}`)
        }
    }

    /** 获取配置数据 */
    public getConfig(): PoetryConfig {
        if (!this.config) {
            this.config = this.loadConfigFile()
        }
        return this.config
    }

    /** 获取当前配置 */
    public getCurrentConfig(): PoetryConfig {
        return this.getConfig()
    }

    /** 更新诗歌编号并保存配置 */
    public async updatePoemNumber(): Promise<void> {
        if (!this.config) {
            this.config = this.loadConfigFile()
        }

        // 更新诗歌编号
        this.config.poemNumber += 1

        // 保存到文件
        try {
            const { writeFileSync } = await import('node:fs')
            const configPath = join(this.configDir, 'config.json')
            writeFileSync(configPath, JSON.stringify(this.config, null, 2), 'utf-8')
        } catch (error) {
            console.warn('保存配置文件失败:', error)
            // 不抛出错误，因为这不应该阻止诗歌生成
        }
    }

    /** 获取名词数据 */
    public getNouns(): WordRecord[] {
        if (!this.nouns) {
            this.nouns = this.loadJsonFile<WordRecord[]>('nouns.json')
        }
        return this.nouns
    }

    /** 获取形容词数据 */
    public getAdjectives(): WordRecord[] {
        if (!this.adjectives) {
            this.adjectives = this.loadJsonFile<WordRecord[]>('adjectives.json')
        }
        return this.adjectives
    }

    /** 获取不及物动词数据 */
    public getIntransitiveVerbs(): WordRecord[] {
        if (!this.intransitiveVerbs) {
            this.intransitiveVerbs = this.loadJsonFile<WordRecord[]>('intransitive-verbs.json')
        }
        return this.intransitiveVerbs
    }

    /** 获取及物动词数据 */
    public getTransitiveVerbs(): WordRecord[] {
        if (!this.transitiveVerbs) {
            this.transitiveVerbs = this.loadJsonFile<WordRecord[]>('transitive-verbs.json')
        }
        return this.transitiveVerbs
    }

    /** 获取叹词数据 */
    public getInterjections(): WordRecord[] {
        if (!this.interjections) {
            this.interjections = this.loadJsonFile<WordRecord[]>('interjections.json')
        }
        return this.interjections
    }

    /** 获取句型结构数据 */
    public getSentenceStructures(): SentenceStructure[] {
        if (!this.sentenceStructures) {
            this.sentenceStructures = this.loadJsonFile<SentenceStructure[]>('sentence-structures.json')
        }
        return this.sentenceStructures
    }

    /** 获取工作结构数据 */
    public getWorkingStructures(): WorkingStructure[] {
        if (!this.workingStructures) {
            this.workingStructures = this.loadJsonFile<WorkingStructure[]>('working-structures.json')
        }
        return this.workingStructures
    }

    /** 获取特殊词数据 */
    public getSpecialWords(): SpecialWord[] {
        if (!this.specialWords) {
            this.specialWords = this.loadJsonFile<SpecialWord[]>('special-words.json')
        }
        return this.specialWords
    }

    /** 获取诗句数据 */
    public getPoemLines(): PoemLine[] {
        if (!this.poemLines) {
            this.poemLines = this.loadJsonFile<PoemLine[]>('poem-lines.json')
        }
        return this.poemLines
    }

    /** 根据词性获取对应的词汇数据 */
    public getWordsByPartOfSpeech(partOfSpeech: string): WordRecord[] {
        switch (partOfSpeech) {
            case 'MM':
            case 'MC':
            case 'MR':
                return this.getNouns()
            case 'DD':
            case 'DI':
            case 'DV':
            case 'DO':
                return this.getIntransitiveVerbs()
            case 'DJ':
                return this.getTransitiveVerbs()
            case 'XA':
                return this.getAdjectives()
            case 'TT':
                return this.getInterjections()
            case 'SS':
                return this.getSpecialWords().map((sw) => ({
                    word: sw.content,
                    vowel: '',
                    property: '',
                }))
            default:
                throw new Error(`未支持的词性: ${partOfSpeech}`)
        }
    }

    /** 重置缓存 */
    public clearCache(): void {
        this.config = null
        this.nouns = null
        this.adjectives = null
        this.intransitiveVerbs = null
        this.transitiveVerbs = null
        this.interjections = null
        this.sentenceStructures = null
        this.workingStructures = null
        this.specialWords = null
        this.poemLines = null
    }
}
