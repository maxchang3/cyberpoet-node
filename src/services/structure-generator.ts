import type { PoeticStyle, RhymeScheme, SentenceStructure, WorkingStructure } from '@/types'
import type { DataService } from './data-service'

/**
 * 句型结构生成器 - 负责创建诗歌的句型结构
 * 还原自原始 Visual FoxPro 的 create_structure 过程
 */
export class StructureGenerator {
    private dataService: DataService

    constructor(dataService: DataService) {
        this.dataService = dataService
    }

    /**
     * 创建诗歌结构
     * @param paragraphCount 段数
     * @param linesPerParagraph 每段行数
     * @param style 诗歌风格
     * @param rhymeScheme 韵脚（可选）
     * @returns 工作结构数组
     */
    public createStructure(
        paragraphCount: number,
        linesPerParagraph: number,
        style: PoeticStyle,
        rhymeScheme?: RhymeScheme
    ): WorkingStructure[] {
        const sentenceStructures = this.dataService.getSentenceStructures()
        const structureCount = sentenceStructures.length

        // 生成临时结构
        const tempStructures: SentenceStructure[] = []

        for (let lineIndex = 1; lineIndex <= linesPerParagraph; lineIndex++) {
            const needsRhyme = this.shouldLineRhyme(lineIndex, linesPerParagraph)
            let selectedStructure: SentenceStructure

            if (style === 'quiet') {
                // 宁静风格：选择原始结构数小于10的句型
                selectedStructure = this.selectQuietStructure(sentenceStructures, needsRhyme, rhymeScheme)
            } else {
                // 奔放风格：可以选择任意句型
                selectedStructure = this.selectBoldStructure(sentenceStructures, needsRhyme, rhymeScheme)
            }

            tempStructures.push(selectedStructure)
        }

        // 为每个段落复制结构并处理成工作结构
        const workingStructures: WorkingStructure[] = []

        for (let paragraphIndex = 1; paragraphIndex <= paragraphCount; paragraphIndex++) {
            for (let lineIndex = 1; lineIndex <= linesPerParagraph; lineIndex++) {
                const baseStructure = tempStructures[lineIndex - 1]
                const needsRhyme = this.shouldLineRhyme(paragraphIndex, paragraphCount)

                // 处理结构元素，扩展复合结构
                const processedStructure = this.processStructureElements(baseStructure, needsRhyme, rhymeScheme)

                workingStructures.push(processedStructure)
            }
        }

        return workingStructures
    }

    /** 判断是否需要押韵 */
    private shouldLineRhyme(lineIndex: number, totalLines: number): boolean {
        if (totalLines % 2 === 0) {
            // 偶数行：第1行和偶数行押韵
            return lineIndex === 1 || lineIndex % 2 === 0
        }
        // 奇数行：奇数行押韵
        return (lineIndex + 1) % 2 === 0
    }

    /** 选择宁静风格的句型结构 */
    private selectQuietStructure(
        structures: SentenceStructure[],
        needsRhyme: boolean,
        rhymeScheme?: RhymeScheme
    ): SentenceStructure {
        let attempts = 0
        const maxAttempts = 100 // 防止无限循环

        while (attempts < maxAttempts) {
            const randomIndex = Math.floor(Math.random() * structures.length)
            const structure = structures[randomIndex]

            // 宁静风格：不选择原始结构数大于等于10的句型
            const isQuiet = structure.elements[9] === '' // J10字段为空表示结构数小于10

            if (!isQuiet) {
                attempts++
                continue
            }

            // 检查韵脚要求
            if (needsRhyme && rhymeScheme && structure.compoundStructureCount === 0) {
                if (structure.limitedRhyme !== rhymeScheme && structure.limitedRhyme !== 'E') {
                    attempts++
                    continue
                }
            }

            return structure
        }

        // 如果无法找到合适的结构，返回第一个结构
        return structures[0]
    }

    /** 选择奔放风格的句型结构 */
    private selectBoldStructure(
        structures: SentenceStructure[],
        needsRhyme: boolean,
        rhymeScheme?: RhymeScheme
    ): SentenceStructure {
        let attempts = 0
        const maxAttempts = 100 // 防止无限循环

        while (attempts < maxAttempts) {
            const randomIndex = Math.floor(Math.random() * structures.length)
            const structure = structures[randomIndex]

            // 检查韵脚要求
            if (needsRhyme && rhymeScheme && structure.compoundStructureCount === 0) {
                if (structure.limitedRhyme !== rhymeScheme && structure.limitedRhyme !== 'E') {
                    attempts++
                    continue
                }
            }

            return structure
        }

        // 如果无法找到合适的结构，返回第一个结构
        return structures[0]
    }

    /** 处理结构元素，扩展复合结构 */
    private processStructureElements(
        structure: SentenceStructure,
        needsRhyme: boolean,
        rhymeScheme?: RhymeScheme
    ): WorkingStructure {
        const processedElements: string[] = []

        for (let i = 0; i < 27; i++) {
            const element = structure.elements[i] || ''

            if (element === '') {
                break
            }

            // 处理特殊的复合结构
            if (element === 'Dd' && structure.limitedRhyme === 'E') {
                // 扩展为 DV+着+DO 结构
                processedElements.push('DV', '着', 'DO')
            } else if (element.length === 2 && element[0] === element[0].toLowerCase()) {
                // 处理小写开头的元素（如：Mm）
                processedElements.push(element.toUpperCase())
            } else {
                processedElements.push(element)
            }
        }

        // 补足到30个元素
        while (processedElements.length < 30) {
            processedElements.push('')
        }

        return {
            compoundStructureCount: structure.compoundStructureCount,
            punctuation: structure.punctuation,
            elements: processedElements,
        }
    }

    /** 随机选择句型结构 */
    private getRandomStructure(structures: SentenceStructure[]): SentenceStructure {
        const randomIndex = Math.floor(Math.random() * structures.length)
        return structures[randomIndex]
    }
}
