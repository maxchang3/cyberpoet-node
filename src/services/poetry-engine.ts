import type {
    GeneratedPoem,
    PartOfSpeech,
    PoetryGenerationOptions,
    WordSelectionContext,
    WorkingStructure,
} from '@/types'
import { DataService } from './data-service'
import { StructureGenerator } from './structure-generator'
import { WordSelector } from './word-selector'

/**
 * 诗歌生成引擎 - 核心诗歌生成逻辑
 * 还原自原始Visual FoxPro的cpzh和cpzhy程序逻辑
 */
export class PoetryEngine {
    private dataService: DataService
    private wordSelector: WordSelector
    private structureGenerator: StructureGenerator

    constructor() {
        this.dataService = DataService.getInstance()
        this.wordSelector = new WordSelector(this.dataService)
        this.structureGenerator = new StructureGenerator(this.dataService)
    }

    /**
     * 生成诗歌
     * @param options 生成选项
     * @returns 生成的诗歌
     */
    public generatePoetry(options: PoetryGenerationOptions): GeneratedPoem {
        console.log(
            `开始生成诗歌，风格：${options.style}，段数：${options.paragraphCount}，每段行数：${options.linesPerParagraph}`
        )

        // 创建诗歌结构
        const structures = this.structureGenerator.createStructure(
            options.paragraphCount,
            options.linesPerParagraph,
            options.style,
            options.rhymeScheme
        )

        // 生成诗句
        const lines: string[] = []
        let lineNumber = 1

        for (const structure of structures) {
            const line = this.generateLine(structure, options)
            lines.push(line)
            lineNumber++
        }

        console.log('诗歌生成完成！')

        return {
            lines,
            options,
            createdAt: new Date(),
        }
    }

    /**
     * 生成单行诗句
     * @param structure 工作结构
     * @param options 生成选项
     * @returns 生成的诗句
     */
    private generateLine(structure: WorkingStructure, options: PoetryGenerationOptions): string {
        let line = ''
        let elementIndex = 0

        while (elementIndex < 27) {
            const element = structure.elements[elementIndex] || ''

            if (element === '') {
                break
            }

            // 判断是否为词性标记
            if (this.isPartOfSpeech(element)) {
                // 规范化词性标记
                const partOfSpeech = this.normalizePartOfSpeech(element)
                const word = this.selectWordForElement(partOfSpeech, options)
                line += word

                // 处理特殊的复合动词结构
                if (partOfSpeech === 'DV' || partOfSpeech === 'DO') {
                    elementIndex += 2 // 跳过后续的修饰符
                }
            } else {
                // 直接添加标点符号或修饰符
                line += element
            }

            elementIndex++
        }

        // 添加句尾标点
        line += structure.punctuation

        return line.trim()
    }

    /**
     * 为结构元素选择词汇
     * @param partOfSpeech 词性
     * @param options 生成选项
     * @returns 选择的词汇
     */
    private selectWordForElement(partOfSpeech: PartOfSpeech, options: PoetryGenerationOptions): string {
        const context: WordSelectionContext = {
            partOfSpeech,
            needsRhyme: options.useRhyme,
            rhymeScheme: options.rhymeScheme,
        }

        return this.wordSelector.selectWord(context)
    }

    /**
     * 将遗留的词性标记规范化为标准格式
     * @param element 原始元素
     * @returns 规范化的词性标记
     */
    private normalizePartOfSpeech(element: string): PartOfSpeech {
        // 处理遗留格式映射
        const legacyMappings: Record<string, PartOfSpeech> = {
            XX: 'XA', // 形容词
            Mm: 'MM', // 名词
            Mr: 'MR', // 人物名词
            Mc: 'MC', // 地点名词
            DB: 'DD', // 动词（暂时映射为不及物动词）
            Dd: 'DD', // 不及物动词
            Dv: 'DV', // 进行时动词
            Do: 'DO', // 结果补语动词
        }

        // 如果是遗留格式，进行转换
        if (element in legacyMappings) {
            return legacyMappings[element]
        }

        // 处理大小写混合的标准格式（如 Mm -> MM）
        const upperElement = element.toUpperCase()
        const partOfSpeechTags: PartOfSpeech[] = ['MM', 'MC', 'MR', 'DD', 'DI', 'DV', 'DO', 'DJ', 'XA', 'TT', 'SS']
        if (partOfSpeechTags.includes(upperElement as PartOfSpeech)) {
            return upperElement as PartOfSpeech
        }

        // 如果已经是标准格式，直接返回
        if (partOfSpeechTags.includes(element as PartOfSpeech)) {
            return element as PartOfSpeech
        }

        // 默认返回名词
        console.warn(`未知的词性标记: ${element}，默认使用名词 (MM)`)
        return 'MM'
    }

    /**
     * 判断字符串是否为词性标记
     * @param element 元素字符串
     * @returns 是否为词性标记
     */
    private isPartOfSpeech(element: string): boolean {
        const partOfSpeechTags = ['MM', 'MC', 'MR', 'DD', 'DI', 'DV', 'DO', 'DJ', 'XA', 'TT', 'SS']

        // 直接匹配标准的词性标记
        if (partOfSpeechTags.includes(element)) {
            return true
        }

        // 处理大小写混合的变体（如 Mm, Mr, Xa 等）
        const upperElement = element.toUpperCase()
        if (partOfSpeechTags.includes(upperElement)) {
            return true
        }

        // 处理特殊的遗留格式
        const legacyMappings: Record<string, boolean> = {
            XX: true, // 可能表示形容词 (XA)
            Mm: true, // 可能表示名词 (MM)
            Mr: true, // 可能表示人物名词 (MR)
            Mc: true, // 可能表示地点名词 (MC)
            DB: true, // 可能表示动词
            Dd: true, // 可能表示动词
            Dv: true, // 可能表示进行时动词
            Do: true, // 可能表示结果补语动词
        }

        return element in legacyMappings
    }

    /**
     * 保存诗歌到文件
     * @param poem 生成的诗歌
     * @param title 诗歌标题
     * @param outputPath 输出路径（可选）
     */
    public async savePoetry(poem: GeneratedPoem, title: string, outputPath?: string): Promise<string> {
        const config = this.dataService.getCurrentConfig()
        const poemNumber = config.poemNumber.toString().padStart(6, '0')
        const filename = `cp${poemNumber}.txt`

        let content = ''
        content += '**************************************\n'
        content += `${title}\n`
        content += `（作品第${poemNumber}号）\n`
        content += '**************************************\n'

        for (const line of poem.lines) {
            content += `${line}\n`
        }

        if (outputPath) {
            const { writeFileSync } = await import('node:fs')
            const { join } = await import('node:path')
            const fullPath = join(outputPath, filename)
            writeFileSync(fullPath, content, 'utf-8')

            // 更新配置中的诗歌编号
            await this.dataService.updatePoemNumber()

            return fullPath
        }

        return content
    }

    /**
     * 获取诗歌年龄信息（从原始程序的计算逻辑迁移）
     * @returns 年龄描述字符串
     */
    public getPoetAge(): string {
        const birthDate = new Date(2002, 3, 4) // 2002年4月4日
        const now = new Date()

        const ageYears = now.getFullYear() - birthDate.getFullYear()
        const ageMonths = now.getMonth() - birthDate.getMonth()
        const ageDays = now.getDate() - birthDate.getDate()

        if (ageYears !== 0) {
            return `${ageYears}岁的计算机诗人火鸟将为您歌唱！`
        }
        if (ageMonths !== 0) {
            return `${ageMonths}个月大的计算机诗人火鸟将为您歌唱！`
        }
        return `${ageDays}天大的计算机诗人火鸟将为您歌唱！`
    }

    /**
     * 获取程序版本信息
     * @returns 版本信息字符串
     */
    public getVersionInfo(): string {
        return `
≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌

　　听吧！
　　${this.getPoetAge()}
　　　　　　　　　　　　　　　　　　Version 2.0 (Node.js)
生日：2002年4月4日　　　　　　　　┌────────────────────────────────────┐
星座：白羊座　　　　　　　　　　　│原作：刘慈欣　　　　　　　　　　　　│
籍贯：山西平定娘子关　　　　　　　│改编：诸葛恒　　　　　　　　　　　　│
爱好：写诗　　　　　　　　　　　　│迁移：Github Copilot Claude Sonnet 4│
　　　　　　　　　　　　　　　　　└────────────────────────────────────┘
　　　　　　　　　　　　　最后更新于 2025 年 6 月 10 日

≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌≌
`
    }
}
