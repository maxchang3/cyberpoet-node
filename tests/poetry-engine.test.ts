import { PoetryEngine } from '@/services/poetry-engine'
import type { PoetryGenerationOptions } from '@/types'
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'

describe('PoetryEngine', () => {
    let engine: PoetryEngine

    beforeEach(() => {
        vi.useFakeTimers()
        engine = new PoetryEngine()
    })

    afterEach(() => {
        vi.useRealTimers()
    })

    describe('诗歌生成功能', () => {
        it('应该能够生成基本诗歌', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 4,
                useRhyme: false,
            }

            const poem = engine.generatePoetry(options)

            expect(poem).toBeDefined()
            expect(poem.lines).toBeDefined()
            expect(Array.isArray(poem.lines)).toBe(true)
            expect(poem.lines.length).toBe(4)
            expect(poem.options).toEqual(options)
            expect(poem.createdAt).toBeInstanceOf(Date)

            // 检查每行都不为空
            poem.lines.forEach((line: string) => {
                expect(typeof line).toBe('string')
                expect(line.trim().length).toBeGreaterThan(0)
            })
        })

        it('应该能够生成宁静风格的诗歌', () => {
            const options: PoetryGenerationOptions = {
                style: 'quiet',
                paragraphCount: 1,
                linesPerParagraph: 2,
                useRhyme: false,
            }

            const poem = engine.generatePoetry(options)

            expect(poem).toBeDefined()
            expect(poem.lines.length).toBe(2)
            expect(poem.options.style).toBe('quiet')
        })

        it('应该能够生成奔放风格的诗歌', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 3,
                useRhyme: false,
            }

            const poem = engine.generatePoetry(options)

            expect(poem).toBeDefined()
            expect(poem.lines.length).toBe(3)
            expect(poem.options.style).toBe('bold')
        })

        it('应该能够生成多段诗歌', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 3,
                linesPerParagraph: 2,
                useRhyme: false,
            }

            const poem = engine.generatePoetry(options)

            expect(poem).toBeDefined()
            expect(poem.lines.length).toBe(6) // 3段 × 2行
        })

        it('应该能够生成押韵诗歌', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 4,
                useRhyme: true,
                rhymeScheme: 'a',
            }

            const poem = engine.generatePoetry(options)

            expect(poem).toBeDefined()
            expect(poem.lines.length).toBe(4)
            expect(poem.options.useRhyme).toBe(true)
            expect(poem.options.rhymeScheme).toBe('a')
        })
    })

    describe('诗歌保存功能', () => {
        it('应该能够生成保存格式的内容', async () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 2,
                useRhyme: false,
            }

            const poem = engine.generatePoetry(options)
            const content = await engine.savePoetry(poem, '测试诗歌')

            expect(typeof content).toBe('string')
            expect(content.includes('测试诗歌')).toBe(true)
            expect(content.includes('**************************************')).toBe(true)
            expect(content.includes('作品第')).toBe(true)

            // 检查诗句是否包含在内容中
            poem.lines.forEach((line: string) => {
                expect(content.includes(line)).toBe(true)
            })
        })

        it('保存内容应该包含正确的编号格式', async () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 1,
                useRhyme: false,
            }

            const poem = engine.generatePoetry(options)
            const content = await engine.savePoetry(poem, '编号测试')

            // 检查编号格式（应该是6位数）
            const numberMatch = content.match(/作品第(\d{6})号/)
            expect(numberMatch).toBeTruthy()
            if (numberMatch) {
                expect(numberMatch[1].length).toBe(6)
            }
        })
    })

    describe('版本信息功能', () => {
        it('应该能够获取版本信息', () => {
            const versionInfo = engine.getVersionInfo()

            expect(typeof versionInfo).toBe('string')
            expect(versionInfo.includes('听吧！')).toBe(true)
            expect(versionInfo.includes('计算机诗人火鸟')).toBe(true)
            expect(versionInfo.includes('Version 2.0')).toBe(true)
            expect(versionInfo.includes('Node.js')).toBe(true)
            expect(versionInfo.includes('刘慈欣')).toBe(true)
            expect(versionInfo.includes('诸葛恒')).toBe(true)
        })

        it('应该能够获取诗人年龄信息', () => {
            // Mock 一个固定的时间点进行测试
            const mockDate = new Date('2025-06-10T12:00:00.000Z')
            vi.setSystemTime(mockDate)

            const ageInfo = engine.getPoetAge()

            expect(typeof ageInfo).toBe('string')
            expect(ageInfo.includes('计算机诗人火鸟将为您歌唱')).toBe(true)

            // 使用快照测试确保输出一致性
            expect(ageInfo).toMatchSnapshot()
        })

        it('应该能够正确计算诗人的月龄（不足一岁时）', () => {
            // Mock 到诗人出生后几个月
            const mockDate = new Date('2002-08-15T12:00:00.000Z')
            vi.setSystemTime(mockDate)

            const ageInfo = engine.getPoetAge()

            expect(typeof ageInfo).toBe('string')
            expect(ageInfo.includes('个月大的计算机诗人火鸟将为您歌唱')).toBe(true)
            expect(ageInfo).toMatchSnapshot()
        })

        it('应该能够正确计算诗人的天数（不足一个月时）', () => {
            // Mock 到诗人出生后几天
            const mockDate = new Date('2002-04-20T12:00:00.000Z')
            vi.setSystemTime(mockDate)

            const ageInfo = engine.getPoetAge()

            expect(typeof ageInfo).toBe('string')
            expect(ageInfo.includes('天大的计算机诗人火鸟将为您歌唱')).toBe(true)
            expect(ageInfo).toMatchSnapshot()
        })
    })

    describe('随机性和一致性测试', () => {
        it('相同参数应该能够生成不同的诗歌（随机性）', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 2,
                useRhyme: false,
            }

            const poems: string[] = []

            // 生成多首诗
            for (let i = 0; i < 5; i++) {
                const poem = engine.generatePoetry(options)
                poems.push(poem.lines.join(''))
            }

            // 应该有不同的结果
            const uniquePoems = new Set(poems)
            expect(uniquePoems.size).toBeGreaterThan(1)
        })

        it('生成的诗歌应该符合基本的中文格式', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 3,
                useRhyme: false,
            }

            const poem = engine.generatePoetry(options)

            poem.lines.forEach((line: string) => {
                // 检查不为空
                expect(line.trim().length).toBeGreaterThan(0)

                // 检查不包含异常字符
                expect(line).not.toMatch(/undefined|null|NaN/)

                // 检查长度合理（中文诗句一般不会太长）
                expect(line.length).toBeLessThan(100)
            })
        })
    })

    describe('错误处理', () => {
        it('应该能够处理极端参数', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 1,
                useRhyme: false,
            }

            expect(() => {
                engine.generatePoetry(options)
            }).not.toThrow()
        })

        it('应该能够处理押韵参数但没有指定韵脚的情况', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 1,
                linesPerParagraph: 2,
                useRhyme: true,
                // 没有指定 rhymeScheme
            }

            expect(() => {
                engine.generatePoetry(options)
            }).not.toThrow()
        })
    })

    describe('性能测试', () => {
        it('应该能够在合理时间内生成诗歌', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 2,
                linesPerParagraph: 4,
                useRhyme: false,
            }

            const startTime = Date.now()
            const poem = engine.generatePoetry(options)
            const endTime = Date.now()

            expect(poem).toBeDefined()
            expect(endTime - startTime).toBeLessThan(5000) // 应该在5秒内完成
        })

        it('生成大型诗歌不应该导致内存问题', () => {
            const options: PoetryGenerationOptions = {
                style: 'bold',
                paragraphCount: 5,
                linesPerParagraph: 8,
                useRhyme: false,
            }

            expect(() => {
                const poem = engine.generatePoetry(options)
                expect(poem.lines.length).toBe(40) // 5段 × 8行
            }).not.toThrow()
        })
    })
})
