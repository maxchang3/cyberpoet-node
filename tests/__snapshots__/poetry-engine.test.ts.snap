// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PoetryEngine > 版本信息功能 > 应该能够正确计算诗人的天数（不足一个月时） 1`] = `"16天大的计算机诗人火鸟将为您歌唱！"`;

exports[`PoetryEngine > 版本信息功能 > 应该能够正确计算诗人的月龄（不足一岁时） 1`] = `"4个月大的计算机诗人火鸟将为您歌唱！"`;

exports[`PoetryEngine > 版本信息功能 > 应该能够获取诗人年龄信息 1`] = `"23岁的计算机诗人火鸟将为您歌唱！"`;
